### ISSUE TITLE
Vert.X Gateway - Error 500 "Policy not found" with custom pluginRepositories

### ISSUE THREAD

**Issue by jhauray** [apiman/apiman#732]
On our deployment we pull Apiman plugins from our Nexus, as a Maven repository.

We ares already using this parameter in apiman.properties file for WildFly gateways :
```yaml
apiman.plugins.repositories=https://<hostname>/nexus/repository/public/
```
=> It works. When an api whose using a plugin policy is call, plugin is downloaded and call successful.

We are tying to replace Wildfly gateways by Vert.X gateways. So, we are settings the same parameters in conf-es.json :
```json
"plugin-registry": {
    "class": "io.apiman.gateway.platforms.vertx3.engine.VertxPluginRegistry",
    "config": {
      "pluginRepositories": ["https://<hostname>/nexus/repository/public/" ],
      "pluginsDir" : "/home/gwvertx/tmp/plugins"
    }
  },
```

=> Butin this case, each api call finish with the following error :
```
{"responseCode":500,"message":"Policy not found: plugin:io.apiman.plugins:apiman-plugins-jwt-policy:1.5.1.Final:war/io.apiman.plugins.jwt.JWTPolicy","trace":null}
```

How to reproduce :
- Deploy a Vert.X gateway 1.3.1 Final.
- Customise "pluginRepositories" parameter in json config, with a Nexus repository.
- Publish an api using a policy provided by a plugin (ex: jwt).
- Send a request to the api.
- See the "Policy not found" error.

Possible origin :
- [VertxPluginRegistry](https://github.com/apiman/apiman/blob/apiman-1.3.1.Final/gateway/platforms/vertx3/vertx3/src/main/java/io/apiman/gateway/platforms/vertx3/engine/VertxPluginRegistry.java) class seem to not consider  "pluginRepositories" parameter in his constructor.
- The code hasn't change since version 1.3.1. So, I think is always present.

**Comment by EricWittmann** [comment]
@msavy Any quick thoughts on this (off the top of your head)?

**Comment by jhauray** [comment]
This problem seems to have already been follow on Jira : https://issues.jboss.org/browse/APIMAN-1301

**Comment by EricWittmann** [comment]
So it looks like we're going to have to dive into the code on this one.  Perhaps it's just a bug that needs to be squashed.

**Comment by jhauray** [comment]
@EricWittmann , I've find the problem in code.

[VertxPluginRegistry](https://github.com/apiman/apiman/blob/apiman-1.3.1.Final/gateway/platforms/vertx3/vertx3/src/main/java/io/apiman/gateway/platforms/vertx3/engine/VertxPluginRegistry.java) constructor instanciate super class [DefaultPluginRegistry](https://github.com/apiman/apiman/blob/apiman-1.3.1.Final/gateway/engine/core/src/main/java/io/apiman/gateway/engine/impl/DefaultPluginRegistry.java), with this constructor :
```java
DefaultPluginRegistry(File pluginsDir, Set<URI> pluginRepositories)
```
And set pluginRepositories with `PluginUtils.getDefaultMavenRepositories()`. But not with values from Map<> config (json conf file).

That's why **pluginRepositories** parameter is not taked in account.

I see two ways to fix the problem :
- limited impact : Calling  `DefaultPluginRegistry.getConfiguredPluginRepositories()` in VertxPluginRegistry constructor, before calling DefaultPluginRegistry constructor  `DefaultPluginRegistry(File pluginsDir, Set<URI> pluginRepositories)`.
- Larger impact : Directly calling `DefaultPluginRegistry(Map<String, String> configMap)` constructor in VertxPluginRegistry constructor. This solution fixe a other problem : conf-es.json **PluginsDir** parameter is not taked in account by Vert.X gateway.

Last point : The problem is always present in master branch.

**Comment by EricWittmann** [comment]
Looking at the code, it seems that option (2) seems pretty low risk.  It looks like it will do exactly what we want.

**Comment by jhauray** [comment]
I agree with you.

Just note that with option 2, if **pluginDir** parameter is not defined,  `VertxPluginRegistry.getTempPluginsDir()` will be no more used. We fall back on `DefaultPluginRegistry.createTempPluginsDir()` method. Temporary plugin directory will be managed in the same way for Vert.X and Wildfly distro. I just hope that the use of a different method was not a deliberate choice to solve an other issue.

**Comment by EricWittmann** [comment]
Oh yes I see - I didn't notice that at first.  The vertx temp plugins directory is a little different than the default one.  Looking at the implementations, I think the vertx one is probably better - so perhaps we could replace the one in `DefaultPluginRegistry` with the one in `VertxPluginRegistry`.  My guess is Marc didn't want to disrupt the original one when we wrote the vertx impl.

**Comment by jhauray** [comment]
Hi @EricWittmann !!

If you agree, I'm up to suggest a merge request.

Just one question : Do you know if it already exist some unit test for VertX plugins loading ?

**Comment by EricWittmann** [comment]
I'm sorry but I don't know offhand.

**Comment by jhauray** [comment]
@msavy, @EricWittmann,  I need your help.

I try to fix this bug. But at testing, I encounter an Exception at VertxPluginRegistry loading : There is no constructor with a Map parameter.

I'm not sure to understand well the loading path for VertX pluginRegistry. For me, call stack seems to be : InitVerticle => ApimanVerticleBase => ApimanVerticleWithEngine => AbstractEngineFactory =>DefaultEngineFactory =>ConfigDrivenEngineFactory.createPluginRegistry().

Is it the right path ?

**Comment by msavy** [comment]
That sounds right off the top of my head

 Did you try adding a constructor with Map<string, string> where it was
complaining?

Can you share a stack trace perhaps

**Comment by jhauray** [comment]
@msavy 

To be able to easily test PluginRegistry instanciation with original class, but without loading a full engine, I write a short extend class :
```java
public class TestConfigDrivenEngineFactory extends ConfigDrivenEngineFactory {

    public TestConfigDrivenEngineFactory(IEngineConfig engineConfig) {
        super(engineConfig);
    }

    public IPluginRegistry createPluginRegistry(){ return super.createPluginRegistry(); }
}
```
The unit test :
```java
@Test
public void CustomizedPluginRegistry(){

        String jsonStr = "{\"plugin-registry\": {\n" +
        "    \"class\": \"io.apiman.gateway.platforms.vertx3.engine.VertxPluginRegistry\",\n" +
        "            \"config\": {\n" +
        "                 \"pluginRepositories\": [\"https://repo1.maven.org/maven2/\" ],\n" +
        "                 \"pluginsDir\" : \"/tmp/plugins\"\n" +
        "            }\n" +
        "    }\n" +
        "}";

        JsonObject jsonObject = new JsonObject(jsonStr);

        IEngineConfig config = new VertxEngineConfig(jsonObject);

        TestConfigDrivenEngineFactory ef = new TestConfigDrivenEngineFactory(config);

        IPluginRegistry pluginRegistry = ef.createPluginRegistry();
}
```
The error stack :
```
java.lang.RuntimeException: java.lang.InstantiationException: io.apiman.gateway.platforms.vertx3.engine.VertxPluginRegistry

	at io.apiman.gateway.engine.impl.ConfigDrivenEngineFactory.instantiate(ConfigDrivenEngineFactory.java:242)
	at io.apiman.gateway.engine.impl.ConfigDrivenEngineFactory.doInstantiate(ConfigDrivenEngineFactory.java:221)
	at io.apiman.gateway.engine.impl.ConfigDrivenEngineFactory.create(ConfigDrivenEngineFactory.java:181)
	at io.apiman.gateway.engine.impl.ConfigDrivenEngineFactory.createPluginRegistry(ConfigDrivenEngineFactory.java:69)
	at io.apiman.gateway.platforms.vertx3.engine.TestConfigDrivenEngineFactory.createPluginRegistry(TestConfigDrivenEngineFactory.java:13)
	at io.apiman.gateway.platforms.vertx3.engine.VertxPluginRegistryTest.CustomizedPluginRegistry(VertxPluginRegistryTest.java:68)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:47)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:44)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:271)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:70)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:50)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:238)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:63)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:236)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:53)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:229)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:309)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:160)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
Caused by: java.lang.InstantiationException: io.apiman.gateway.platforms.vertx3.engine.VertxPluginRegistry
	at java.lang.Class.newInstance(Class.java:427)
	at io.apiman.gateway.engine.impl.ConfigDrivenEngineFactory.instantiate(ConfigDrivenEngineFactory.java:240)
	... 27 more
Caused by: java.lang.NoSuchMethodException: io.apiman.gateway.platforms.vertx3.engine.VertxPluginRegistry.<init>()
	at java.lang.Class.getConstructor0(Class.java:3082)
	at java.lang.Class.newInstance(Class.java:412)
	... 28 more
```

And about adding a constructor, it's not possible. In this context, we need Vertx object used by createHttpClient() : [VertxPluginRegistry.java(L64)](https://github.com/apiman/apiman/blob/f29772a1a2bd67f88a10eb724853c3dd6b996466/gateway/platforms/vertx3/vertx3/src/main/java/io/apiman/gateway/platforms/vertx3/engine/VertxPluginRegistry.java#L64)

I see two possible explanations : I didn't find the right loading path or VertxPluginRegistry is not used.

**Comment by msavy** [comment]
It was a long time ago I wrote this stuff, I can't quite remember. I'll
have a look over the weekend if I can.

**Comment by jhauray** [comment]
Thanks!

**Comment by jhauray** [comment]
@msavy ,

I find my mistake : The right class is "VertxConfigDrivenEngineFactory" and not "ConfigDrivenEngineFactory".

Sorry for this.

Have a good weekend!

**Comment by jhauray** [comment]
@EricWittmann,
 
The issue is fixed by PR #788.

It is deploy on our api management staging platform. It works !
